### DISCUSSION TITLE
3D model black on certain Android devices since v1.9.0

### DISCUSSION THREAD

**Discussion by alansegar** [google/model-viewer#3477] [Q&A]
### Description

Astronaut model on modelviewer.dev and different models on other sites are completely black on certain Android devices when upgrading beyond v1.8.0. I.e. it works OK on 1.8.0 or lower but fails on 1.9.0 or higher on the same device / browser.
Console shows error: `model-viewer.js:16340 THREE.WebGLRenderer: EXT_color_buffer_float extension not supported.`
`model-viewer.js:16340 ` (line number and snippet are from a later version - whatever is currently on modelviewer.dev but the error is the same on v1.9.0).

```
		init: function ( capabilities ) {

			if ( capabilities.isWebGL2 ) {
				getExtension( 'EXT_color_buffer_float' );
// it comes into this if, but getExtension returns null. This happens in both v1.8.0 and v1.9.0
			} else {

				getExtension( 'WEBGL_depth_texture' );
				getExtension( 'OES_texture_float' );
				getExtension( 'OES_texture_half_float' );
				getExtension( 'OES_texture_half_float_linear' );
				getExtension( 'OES_standard_derivatives' );
				getExtension( 'OES_element_index_uint' );
				getExtension( 'OES_vertex_array_object' );
				getExtension( 'ANGLE_instanced_arrays' );

			}

			getExtension( 'OES_texture_float_linear' );
			getExtension( 'EXT_color_buffer_half_float' );
			getExtension( 'WEBGL_multisampled_render_to_texture' );

		},

		get: function ( name ) {

			const extension = getExtension( name );

			if ( extension === null ) {
// This is output to console only on v1.9.0+
				console.warn( 'THREE.WebGLRenderer: ' + name + ' extension not supported.' ); 

			}

			return extension;

		}

	};
```
Stacktrace:
```
get (model-viewer.js:16340) 
E (model-viewer.js:22375)  
z (model-viewer.js:23430)
Tr.setupRenderTarget (model-viewer.js:23846)
Or.setRenderTarget (model-viewer.js:27783)
_textureToCubeUV (model-viewer.js:15652)
_fromTexture (model-viewer.js:15471)
fromEquirectangular (model-viewer.js:15366)
get (model-viewer.js:16178)
(anonymous) (model-viewer.js:27365)
Or.renderBufferDirect (model-viewer.js:26607)
Dt (model-viewer.js:27215)
Pt (model-viewer.js:27184)
Lt (model-viewer.js:27106)
Or.render (model-viewer.js:26926)
render (model-viewer.js:55383)
(anonymous) (model-viewer.js:55232)
(anonymous) (model-viewer.js:26799)
r (model-viewer.js:12599)
requestAnimationFrame (async)
r (model-viewer.js:12601)
```
call comes from the following code where `internalFormat` is `34842` on v1.9.0+
I _think_ `internalFormat` is `32856` on v1.8.0 when it works OK but I'm debugging minified code so it's hard to tell for sure that it's the same function.
```
getInternalFormat
....
		if ( internalFormat === 33325 || internalFormat === 33326 ||
			internalFormat === 33327 || internalFormat === 33328 ||
			internalFormat === 34842 || internalFormat === 34836 ) {

			extensions.get( 'EXT_color_buffer_float' );

		}
```
Did the logic in `getInternalFormat` change in 1.9.0?
Is there a way to get the newer versions to work on these devices or is the only solution to stay on 1.8.0?

#### Live Demo
Can be seen on modelviewer.dev with the (later than 1.9.0) version deployed on there today. Screenshot on https://github.com/google/model-viewer/issues/2949 shows how it looks (but that is describing a different issue)

### Version
- model-viewer: v1.9.0 onwards

### Browser Affected

- [x] Chrome, version: 100.0.4896.127
- [ ] Edge
- [ ] Firefox
- [ ] IE
- [ ] Safari

### OS

- [x] Android
- [ ] iOS
- [ ] Linux
- [ ] MacOS
- [ ] Windows

### AR
N/A

**Comment by elalish** [comment]
Yes, this logic changed and went back and forth a bit (have you tested all the versions beyond 1.9?). We upgraded to floating point textures internally for PMREM, which simplified the code and improved the speed quite a bit. We couldn't find any devices that didn't support the extension we needed, but it was pretty likely something existed somewhere. What device(s) have you found this problem on?

**Comment by alansegar** [comment]
I've just tested on 1.11.0 and 1.12.0. It behaves the same as 1.9.0 on both of those. The only device I know of currently with this issue is Huawei P8 (quite an old device).

**Comment by elalish** [comment]
Well, that's good at least. It's just not possible for us to support every old device, though we make a best effort especially for the more common ones. At least the old library versions are always available if you really need it for that device.
